src/librustc_mir/monomorphize/collector.rs

```diff
@@ -729,6 +733,25 @@ fn should_monomorphize_locally<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, instance:
                 true
             }
         }
+    };
+
+    fn is_available_upstream_generic<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                                               def_id: DefId,
+                                               substs: &'tcx Substs<'tcx>)
+                                               -> bool {
+        debug_assert!(!def_id.is_local());
+
+        if !tcx.share_generics() {
+            return false
+        }
+
+        if substs.types().next().is_none() {
```

Comment:
I'm personally not too well versed in generics in rustc, but from that naive point of view I'm not entirely sure what this clause is doing (but the others here make sense to me). Perhaps a comment could be added why this is here?

Reply:
Will do.